plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs

```diff
@@ -0,0 +1,118 @@
+-- | The CEK machine.
+-- Rules are the same as for the CK machine from "Language.PlutusCore.Evaluation.CkMachine",
+-- except we do not use substitution and use environments instead.
+-- The CEK machine relies on variables having
+-- 1. equal 'Unique's whenever they have equal string names
+-- 2. non-equal 'Unique's whenever they have non-equal string names
+-- I.e. 'Unique's are used instead of string names, so a renamer pass is required.
+-- This is for efficiency reasons.
+-- The CEK machine generates booleans along the way which might contain globally non-unique 'Unique's.
+-- This is not a problem as the CEK machines handles name capture by design.
+
+module PlutusCoreInterpreter.CekMachine
+    ( EvaluationResult (..)
+    , evaluateCek
+    , runCek
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.MachineException (MachineError (..), MachineException (..))
+import           Language.PlutusCore.Evaluation.Result           (EvaluationResult (..))
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Data.IntMap                                     (IntMap)
+import qualified Data.IntMap                                     as IntMap
+
+type Plain f = f TyName Name ()
+
+-- | Environments used by the CEK machine.
+-- Each row is a mapping from the 'Unique' representing a variable to the 'Value' the
+-- variable stands for and the environment the value was defined in.
+newtype Environment = Environment (IntMap (Plain Value, Environment))
+
+data Frame
+    = FrameApplyFun Environment (Plain Value)    -- ^ @[V _]@
+    | FrameApplyArg Environment (Plain Term)     -- ^ @[_ N]@
+    | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
+    | FrameUnwrap                                -- ^ @(unwrap _)@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+
+type Context = [Frame]
+
+-- | Extend an environment with a variable name, the value the variable stands for
+-- and the environment the value is defined in.
+extendEnvironment :: Name () -> Plain Value -> Environment -> Environment -> Environment
+extendEnvironment argName arg argEnv (Environment oldEnv) =
+    Environment $ IntMap.insert (unUnique $ nameUnique argName) (arg, argEnv) oldEnv
+
+-- | Look up a name in an environment.
+lookupName :: Name () -> Environment -> Maybe (Plain Value, Environment)
+lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+
+-- | The computing part of the CEK machine.
+computeCek :: Environment -> Context -> Plain Term -> EvaluationResult
```

Comment:
No operators this time? ;)

In all seriousness, I think it would be nice if this was as close to the CEK machine as possible, stylistically too. E.g. they both have a computing and a returning part, and a context, etc.

Reply:
er, the CK machine

Reply:
My preference would then be to use `computeCk` and `returnCk` in the CK machine, but not a big deal.

Reply:
It's more important for a reference interpreter to match the specification it's from rather than subsequent interpreters.

Reply:
In general I agree with the sentiment to try to make the two as similar as possible, but in this case, I think, the overriding priorities ought to be that (1) the CK machine is as close to the LaTeX in the spec as possible (as @effectfully says) and (2) the CEK machine being good production Haskell code. We'll test CEK against CK; so, that is how we'll get them to match up.